### PR TITLE
sunos: handle rearm errors

### LINKED ISSUE #1078: uv__fs_event_rearm return codes are ignored
I ran into a problem where fs.watch would pretend to be watching a file when the port_associate call was actually returning ENOENT. The reason appears to be:

https://github.com/joyent/libuv/blob/6149b66ccb9998f561a207162195bf5cba647280/src/unix/sunos.c#L415

is not checking the return value from uv__fs_event_rearm() even though that should have returned the errno from the failed port_associate().

### PR DESCRIPTION
**Pull request by indutny**
fix #1078

### CODE REVIEW

**Comment by indutny** [comment]
@bcantrill May I ask you to review it?

File: src/unix/sunos.c [hunk 1]
```diff
@@ -361,11 +363,13 @@ static void uv__fs_event_read(uv_loop_t* loop,
     assert(events != 0);
     handle->fd = PORT_FIRED;
     handle->cb(handle, NULL, events, 0);
+
+    if (handle != NULL && handle->fd != PORT_DELETED) {
```
Comment (indutny) [inline src/unix/sunos.c]: It should be a better place to rearm handle if we are going to receive multiple events from port.

**Comment by indutny** [comment]
@bcantrill or @jclulow any idea if I could land it?

File: src/unix/sunos.c (hunk 1 above)
Comment (jclulow) [inline src/unix/sunos.c]: Surely `handle` is _always_ `!= NULL` here, or we would have aborted while dereferencing on line 365?

Comment (indutny) [inline src/unix/sunos.c]: Good catch.

Comment (indutny) [inline src/unix/sunos.c]: @jclulow pushed fix.

File: src/unix/sunos.c [hunk 2]
```diff
@@ -361,11 +363,12 @@ static void uv__fs_event_read(uv_loop_t* loop,
     assert(events != 0);
     handle->fd = PORT_FIRED;
     handle->cb(handle, NULL, events, 0);
+
+    if (handle->fd != PORT_DELETED)
```
Comment (saghul) [inline src/unix/sunos.c:367]: handle->fd was just set to PORT_FIRED above, so I guess this if is redundant?

Comment (saghul) [inline src/unix/sunos.c:367]: Ah, but the user maybe closed the handle in the callback! Forget that.

Comment (indutny) [inline src/unix/sunos.c:367]: It could become `DELETED` inside callback.

File: src/unix/sunos.c [hunk 3]
```diff
@@ -387,14 +390,16 @@ int uv_fs_event_init(uv_loop_t* loop,
...
   handle->fd = PORT_UNUSED;
   handle->cb = cb;
 
   memset(&handle->fo, 0, sizeof handle->fo);
   handle->fo.fo_name = handle->filename;
-  uv__fs_event_rearm(handle);
+  if (uv__fs_event_rearm(handle) != 0)
+    return -1;
+
+  uv__handle_start(handle); /* FIXME shouldn't start automatically */
```
Comment (saghul) [inline src/unix/sunos.c:402]: I changed this a while ago so we now have uv_fsevent_start, I guess we should do that here?

Comment (saghul) [inline src/unix/sunos.c:402]: Sorry, this is the v0.10 branch, disregard that as well.

File: src/unix/sunos.c (hunk 2 above)
Comment (jclulow) [inline src/unix/sunos.c:367]: I would brace the outer `if`, for readability -- but that's just a style nit.

**Comment by jclulow** [comment]
I posted another style nit, but with or without that: LGTM!

**Comment by indutny** [comment]
Thank you!

**Comment by indutny** [comment]
Thank you, landed in 703a9e601e976b74ce459e4db9e23fca107d3f4b

**Pull request closed without merging**